Use initializer list in Arrays

KoBeWi opened the pull request:
Allows to replace
```C++
Array a;
a.append("something");
a.append(1);
```
with
```C++
Array a{ "something", 1 };
```
I applied it to all Arrays I found it applicable.

KoBeWi commented:
I'm having trouble adding initializer list to TypedArray. I did this:
```C++
_FORCE_INLINE_ TypedArray(std::initializer_list<T> p_init) :
		Array(p_init) {
	set_typed(Variant::OBJECT, T::get_class_static(), Variant());
}
```
but it doesn't work. I get `Can't convert from "initializer_list" to "TypedArray<String>"` error (C2440).

AThousandShips commented:
It can't convert it to `std::typed_array<Variant>`

bruvzg commented:
> I'm having trouble adding initializer list to TypedArray. I did this:

`TypedArray` is still using `Vector<Variant>` inside, probably will require using a full loop like `Vector` do to convert each element, something like:

```c++
_FORCE_INLINE_ TypedArray(std::initializer_list<T> p_init) {
	set_typed(Variant::OBJECT, T::get_class_static(), Variant());

	Error err = _p->array.resize(p_init.size());
	ERR_FAIL_COND(err);

	CowData<T>::Size i = 0;
	for (const T &element : p_init) {
		_p->array.set(i++, element);
	}
}
```

AThousandShips commented:
```cpp
_p = memnew(ArrayPrivate);
_p->refcount.init();
```
This part would be unnecessary as it's done in the super constructor

KoBeWi commented:
Shouldn't it be possible to initialize the Vector using Array's constructor? I tried `initializer_list<Variant>` instead of `initializer_list<T>` and I get the same error. Note that the problem is with the signature/call, it won't work even if the constructor body is empty.

What worked is wrapping an Array inside TypedArray:
```C++
TypedArray<String> cs(Array({ ".", ",", "(", "=", "$", "@", "\"", "\'" }));
```
but that's rather ugly.

KoBeWi commented:
There were only a few usages of TypedArray that could use this, so did not add a constructor in the end. The above case can be further simplified to
```C++
TypedArray<String> cs(Array{ ".", ",", "(", "=", "$", "@", "\"", "\'" });
```
which does not look that bad.

KoBeWi committed:
Use initializer list in Arrays

KoBeWi commented on tests/scene/test_text_edit.h:
```diff
@@ -1787,13 +1759,11 @@ TEST_CASE("[SceneTree][TextEdit] text entry") {
-			Array args3;
-			args2.push_back(0);
+			Array args3{ 0 };
 			lines_edited_args.push_front(args2);
```
Apparently this failed a test, because I accidentally fixed wrong code 🤦‍♂️

kitbdev commented:
- Overlaps with https://github.com/godotengine/godot/pull/86015
- Partially addresses https://github.com/godotengine/godot-proposals/issues/1357

In my PR, I also added it for TypedArray and Variant, but I didn't change uses anywhere in the codebase.
Adding it to Variant allows for nested initializer arrays.

I initialized the Array differently, assigning the array to a Vector. I don't know which way is better.
For the TypedArray, I added it only to the macro part since it works if the type is known.

I also added tests, since #50493 also did that, though I don't know how necessary they are.
The test_code_edit.h and test_text_edit.h are cleaned up in #86978. The text edit ones are changed a lot, so we could leave those untouched for now. The code edit ones are mostly the same, but they could be nested.
We should be able to replace the `build_array()` method with initializer lists in `test_array.h`, `test_dictionary.h`, `test_variant.h`, and  `test_navigation_server_3d.h`.

KoBeWi commented:
> I also added tests, since https://github.com/godotengine/godot/pull/50493 also did that, though I don't know how necessary they are.

Well, I added initializer list usage in existing tests, so if they still pass it means it works, I guess.

> Overlaps with https://github.com/godotengine/godot/pull/86015

The biggest part of this PR is updating the codebase to use the new functionality, so your PR could be merged first.

KoBeWi commented:
Rebased. The list initialization was implemented in another PR, so I changed this one to introduce usages.

There is one unexpected thing. This code:
```C++
Array a1;
Array a2 { a1 };
```
makes `a2` a copy of `a1` instead of nesting it. Something to keep in mind.

KoBeWi commented on tests/core/io/test_json_native.h:
```diff
@@ -188,9 +176,7 @@ TEST_CASE("[JSON][Native] Conversion between native and JSON formats") {
-	TypedArray<Resource> res_arr;
-	res_arr.push_back(res);
-
+	TypedArray<Resource> res_arr{ res };
```
This change seems to make the test fail for some reason.

AThousandShips replied:
Should be solved by adding a dedicated constructor to `TypedArray`, this probably goes via `Array` now which then is converted incorrectly

kitbdev replied:
I only added the initializer list for the specialized TypedArrays (`MAKE_TYPED_ARRAY`), so it doesn't work with objects.
This goes through the `TypedArray(const Variant &p_variant)` non-macro constructor and I think it tries to convert the object into an array so it is left empty.
I didn't add it since I was having issues with it, but I was probably doing something wrong.

Adding an initializer list constructor to the non-macro TypedArray should fix it, or I can open a PR for it.
```gdscript
_FORCE_INLINE_ TypedArray(std::initializer_list<Variant> p_init) : TypedArray(Array(p_init)) {}
```

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

Binary size (Linux x86_64 stripped release export template with `production=yes lto=full`):

- `master`: 72,031,960 bytes
- This PR (rebased on top of `master`): 72,023,768 bytes (-8 KB)

tomfull123 commented:
@KoBeWi looks like theres some conflicts to resolve 😊

tomfull123 commented on tests/scene/test_viewport.h:
```diff
@@ -409,10 +409,8 @@ TEST_CASE("[SceneTree][Viewport] Controls and InputEvent handling") {
-				Array signal_args;
-				signal_args.push_back(node_array);
+				Array signal_args{ { node_a } };
+				print_line(signal_args);
```
Is this an accidental debug print?

tomfull123 commented on tests/scene/test_code_edit.h:
```diff
@@ -905,8 +905,7 @@ TEST_CASE("[SceneTree][CodeEdit] delimiters") {
 			/* Set should override existing, and test multiline */
-			TypedArray<String> delimiters;
-			delimiters.push_back("^^ ^^");
+			TypedArray<String> delimiters(Array{ "^^ ^^" });
```
Should this not use the TypedArray initializer list directly?

tomfull123 reviewed (approved):
Looks good to me!

akien-mga commented on core/debugger/debugger_marshalls.cpp:
```diff
@@ -36,8 +36,7 @@
 Array DebuggerMarshalls::ScriptStackDump::serialize() {
-	Array arr;
-	arr.push_back(frames.size() * 3);
+	Array arr{ frames.size() * 3 };
```
Do we want to use this syntax, or this one?
```suggestion
	Array arr = { frames.size() * 3 };
```
I believe so far we've used mostly the latter for initializer lists.

KoBeWi replied:
Are there any functional differences between the two?

tomfull123 replied:
https://stackoverflow.com/questions/63113374/c-vector-initialization-list-vs-assignment
There's a slight difference, not sure if it matters though?

tomfull123 replied:
Reading online it seems most people prefer without the `=`, which is my preference too

KoBeWi replied:
Changed all occurrences, hopefully I didn't miss anything.

Though seems like we use the direct initialization in quite a few places, mostly for non-array types ([including cases like this](https://github.com/godotengine/godot/blob/594d64ec244b6b99321e2096987d4d69de4c8845/platform/macos/export/export_plugin.cpp#L1247)).

tomfull123 commented on editor/debugger/debug_adapter/debug_adapter_protocol.cpp:
```diff
@@ -279,10 +270,7 @@ int DebugAdapterProtocol::parse_variant(const Variant &p_var) {
-			arr.push_back(x.to_json());
-			arr.push_back(y.to_json());
-			arr.push_back(origin.to_json());
+			Array arr = ={ x.to_json(), y.to_json(), origin.to_json() };
```
Looks like you've accidentally put an extra `=`

tomfull123 commented on editor/themes/editor_fonts.cpp:
```diff
@@ -210,13 +210,14 @@ void editor_register_fonts(const Ref<Theme> &p_theme) {
-		emoji_font_names.push_back("Twitter Color Emoji");
-		emoji_font_names.push_back("OpenMoji");
-		emoji_font_names.push_back("EmojiOne Color");
+		PackedStringArray emoji_font_names{
```
Should this use `=` too?

tomfull123 commented on tests/core/io/test_json_native.h:
```diff
@@ -161,25 +161,13 @@ TEST_CASE("[JSON][Native] Conversion between native and JSON formats") {
-	int_arr.push_back(2);
-	int_arr.push_back(3);
-
+	TypedArray<int64_t> int_arr{ 1, 2, 3 };
```
Should this use `=` too?

tomfull123 commented on tests/core/io/test_json_native.h:
```diff
@@ -188,9 +176,7 @@ TEST_CASE("[JSON][Native] Conversion between native and JSON formats") {
-	TypedArray<Resource> res_arr;
-	res_arr.push_back(res);
-
+	TypedArray<Resource> res_arr{ res };
```
Here too

tomfull123 commented on tests/core/math/test_expression.h:
```diff
@@ -321,15 +321,11 @@ TEST_CASE("[Expression] Boolean expressions") {
-	PackedStringArray parameter_names;
-	parameter_names.push_back("foo");
-	parameter_names.push_back("bar");
+	PackedStringArray parameter_names{ "foo", "bar" };
```
And here 😅

tomfull123 commented on tests/core/variant/test_array.h:
```diff
@@ -617,11 +596,8 @@ TEST_CASE("[Array] Iteration and modification") {
-
-	TypedArray<double> a2;
-	a2.push_back(1.0);
+	TypedArray<int> a1{ 1 };
```
and here

tomfull123 commented on tests/core/variant/test_variant_utility.h:
```diff
@@ -92,15 +92,8 @@ TEST_CASE("[VariantUtility] Type conversion") {
-		packed.push_back(3.4);
-		packed.push_back(5.6);
+		Array arr = { 1.2, 3.4, 5.6 };
+		PackedFloat64Array packed{ 1.2, 3.4, 5.6 };
```
and here

tomfull123 commented on tests/scene/test_code_edit.h:
```diff
@@ -905,8 +905,7 @@ TEST_CASE("[SceneTree][CodeEdit] delimiters") {
 			/* Set should override existing, and test multiline */
-			TypedArray<String> delimiters;
-			delimiters.push_back("^^ ^^");
+			TypedArray<String> delimiters(Array{ "^^ ^^" });
```
Are we able to use the initializer list for these variables too?

tomfull123 replied:
I see the discussion above, but I'm not sure if that was before you added the initializer_list constructor and I've seen other instances of TypedArray using the initializer

KoBeWi replied:
I didn't find more of them.

tomfull123 commented on tests/core/io/test_json.h:
```diff
@@ -152,15 +152,7 @@ TEST_CASE("[JSON] Parsing escape sequences") {
-	valid_escapes.push_back("n;\n");
-	valid_escapes.push_back("r;\r");
-	valid_escapes.push_back("t;\t");
+	TypedArray<String> valid_escapes(Array{ "\";\"", "\\;\\", "/;/", "b;\b", "f;\f", "n;\n", "r;\r", "t;\t" });
```
Are we able to use the initializer list for this variable?

tomfull123 commented on tests/scene/test_code_edit.h:
```diff
@@ -3959,7 +3950,7 @@ TEST_CASE("[SceneTree][CodeEdit] completion") {
 
 		Array signal_args;
-		signal_args.push_back(Array());
+		signal_args.append(Array());
```
This should be able to use the initializer too, right?

kitbdev commented:
Also two new ones from https://github.com/godotengine/godot/pull/104573/files#diff-14086706f67239e8decda868cadde74ed52de968211b58cba09c20d966b1e7e8
`editor/debugger/editor_expression_evaluator.cpp` line 77 and `editor/debugger/script_editor_debugger.cpp` line 303.

tomfull123 reviewed (approved):
Looks good to me!

tomfull123 commented:
Whats this PR waiting on? Can it be merged? 😊

akien-mga commented:
Thanks!
